builtin/logical/transit/path_keys.go

```diff
@@ -154,16 +155,16 @@ func (b *backend) pathPolicyWrite(ctx context.Context, req *logical.Request, d *
 		return logical.ErrorResponse(fmt.Sprintf("unknown key type %v", keyType)), logical.ErrInvalidRequest
 	}
 
-	p, lock, upserted, err := b.lm.GetPolicyUpsert(ctx, polReq)
-	if lock != nil {
-		defer lock.RUnlock()
-	}
+	p, upserted, err := b.lm.GetPolicy(ctx, polReq)
 	if err != nil {
 		return nil, err
 	}
 	if p == nil {
 		return nil, fmt.Errorf("error generating key: returned policy was nil")
 	}
+	if b.System().CachingDisabled() {
+		p.Unlock()
```

Comment:
Why in some places, if caching is disabled, it's unlocked, but in other places if caching is _not_ disabled, it's locked but not exclusively? Just curious.

Reply:
If caching is enabled, we return a non-locked Policy. It is then up to the caller to decide whether to lock it exclusively or not depending on the operation. Encryption/decryption don't need an exclusive lock, rotate does, etc.

If caching is disabled, we actually share locks between the lock manager and the policy. This ensures that for a given policy that we are reading from disk (again, no cache), before we read the current value from disk the write lock was grabbed; therefore, if we rotate or in some other way modify the value, by the time we give up the lock (after persisting) we know that no other caller has modified the policy at the same time. So if cache is enabled we have to grab a lock at appropriate places and in these functions we'll always need to unlock. However, if cache is disabled, in some places in the cache-enabled case we actually don't need to grab a lock, but in the cache-disabled case we already have one -- so we need to unlock.

Reply:
That makes sense. Thanks for the explanation.